mixer/v1/config/client/auth.proto

```diff
@@ -0,0 +1,148 @@
+option (gogoproto.gostring_all) = false;
+
+message AuthPolicy {
+  // APIKey defines configuration for generating the API key attribute
+  // from HTTP requests.
+  //
+  // See https://swagger.io/docs/specification/authentication/api-keys
+  // for a general overview of API keys as defined by OpenAPI.
+  message APIKey {
+    oneof key {
+      // API Key is sent as a query parameter. `query` represents the
+      // query string parameter name.
+      //
+      // For example, `query=api_key` should be used with the
+      // following request:
+      //
+      //     GET /something?api_key=abcdef12345
+      //
+      string query = 1;
+      // API key is send in a request header. `header` represents the
+      // header name.
+      //
+      // For example, `header=X-API-KEY` should be used with the
+      // following request:
+      //
+      //     GET /something HTTP/1.1
+      //     X-API-Key: abcdef12345
+      //
+      string header = 2;
+
+      // API key is sent in a
+      // [cookie](https://swagger.io/docs/specification/authentication/cookie-authentication),
+      //
+      // For example, `cookie=X-API-KEY` should be used for the
+      // following request:
+      //
+      //     GET /something HTTP/1.1
+      //     Cookie: X-API-KEY=abcdef12345
+      //
+      string cookie = 3;
+    }
+
+    // Attribute name to use when forwarding api_key to mixer,
+    // e.g. api.key
+    string attribute_name = 4;
+  }
+
+  message JWT {
+    // TODO - does filter need a hint for where to extract JWT token
+    // from, e.g. URL, query parameter, or header?
+
+    // Value to insert into `sec-istio-auth-userinfo` header (or
+    // attribute?)
+    enum UserInfoType {
+      PAYLOAD = 0; // Payload JSON
+      PAYLOAD_BASE64_URL = 1; // base64url-encoded Payload JSON
+      HEADER_PAYLOAD_BASE64_URL = 2; // JWT with signature
+    }
+    UserInfoType user_infoType;
+
+    // Duration after which the cached public key should be
+    // expired. The default value of zero disables caching.
+    google.protobuf.Duration public_key_cache_expiration = 4;
+
+    message Issuer {
+      // Issuer claim value, e.g. "https://accounts.example.com"
+      string name;
+
+      // URI for public key
+      string uri;
+
+      oneof public_key {
+        string pem = 0;
+        string jwks = 1;
+      }
+
+      // service name for public key
+      string cluster;
+
+      // List of allowed audience claims. Any audience will be
+      // accepted if the audience list is empty.
+      repeated string audiences;
+    }
+
+    // List of issuers to verify for this JWT.
+    repeated Issuer issuers;
+  }
+
+  oneof policy {
+    // API key as defined by
+    // https://swagger.io/docs/specification/authentication/api-keys
+    APIKey api_key = 1;
+
+    // JWT as defined by https://tools.ietf.org/html/rfc7519
+    JWT jwt = 2;
+  }
+}
+
+// Determines how to apply auth policies for individual requests.
+message AuthPolicySpec {
```

Comment:
I think api_key extraction and JWT auth should be two separate config spec.

Reply:
I see your point.  I am fine with putting api.key and auth config in the same config.